keras/src/layers/rnn/rnn.py

```diff
@@ -382,6 +386,21 @@ def call(
                 initial_state = self.get_initial_state(
                     batch_size=ops.shape(sequences)[0]
                 )
+        if self.stateful:
+            actual_batch_size = ops.shape(sequences)[0]
+            if (
+                self._expected_batch_size is not None
+                and actual_batch_size is not None
```

Comment:
In the case of a non-constant batch size, this is dynamic and can be something that's not `None` and not an `int`. It can be a `tf.Tensor` for instance.

So instead of `actual_batch_size is not None` please do `not instance(actual_batch_size, int)`.